### PR TITLE
[14.0] shopfloor, delivery: new option 'Process as pre-packaged' products

### PR DESCRIPTION
**Pull request by sebalix**
This option allows to scan product packaging for bulk products and process the packaging qty.

Ref. rau-38

### CODE REVIEW

File: shopfloor/services/delivery.py [hunk 1]
```diff
@@ -112,6 +112,7 @@ def scan_deliver(self, barcode, picking_id=None):
         * deliver: always return here with the data for the last touched
         picking or no picking if the picking has been set to done
         """
+        allow_prepackaged_product = self.work.menu.allow_prepackaged_product
```
Comment (simahawk) [inline shopfloor/services/delivery.py:120]: this should probably come from https://github.com/OCA/wms/blob/14.0/shopfloor_base/services/service.py#L282
but now that I look again at it, we are not overriding the value from the scenario w/ the value from the menu....
If you have time you could maybe have a look at this.

File: shopfloor/services/delivery.py [hunk 2]
```diff
@@ -139,16 +142,34 @@ def scan_deliver(self, barcode, picking_id=None):
             if lot:
                 return self._deliver_lot(picking, lot)
 
+        if not barcode_valid and allow_prepackaged_product:
+            packaging = search.packaging_from_scan(barcode)
+            if packaging:
+                packaging_qty = packaging.product_uom_id._compute_quantity(
```
Comment (simahawk) [inline shopfloor/services/delivery.py:155]: add a comment like "Thus we use the full qty of the packaging" or similar

**Review by simahawk — APPROVED** [review]
LG overall

File: shopfloor/services/delivery.py (hunk 2 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:155]: done, and docstring of the `scan_deliver` method updated as well

File: shopfloor/services/delivery.py (hunk 1 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:120]: I didn't know about this `options` property. I'll check probably later if I have time.

File: shopfloor/actions/search.py [hunk 3]
```diff
@@ -34,12 +34,12 @@ def picking_from_scan(self, barcode):
...
-    def product_from_scan(self, barcode):
+    def product_from_scan(self, barcode, use_packaging=True):
         model = self.env["product.product"]
         if not barcode:
             return model.browse()
         product = model.search([("barcode", "=", barcode)], limit=1)
-        if not product:
+        if not product and use_packaging:
             packaging = self.env["product.packaging"].search(
                 [("product_id", "!=", False), ("barcode", "=", barcode)], limit=1
```
Comment (jbaudoux) [inline shopfloor/actions/search.py]: call packaging_from_scan

File: shopfloor/actions/search.py [hunk 4]
```diff
@@ -58,6 +58,14 @@ def lot_from_scan(self, barcode, products=None, limit=1):
             domain.append(("product_id", "in", products.ids))
         return model.search(domain, limit=limit)
 
+    def packaging_from_scan(self, barcode):
+        model = self.env["product.packaging"]
+        if not barcode:
+            return model.browse()
+        return model.search(
+            [("barcode", "=", barcode), ("product_id", "!=", False)], limit=1
```
Comment (jbaudoux) [inline shopfloor/actions/search.py:64]: and qty > 0 ?

**Review by jbaudoux — CHANGES_REQUESTED** [review]
Reviewing the behavior when scanning a product barcode, so no prepacked product (product.packaging) or no packed product (stock.quant.package), the outcome is quite dangerous as it will deliver completely a matched move line.

We need to be able to disable that product scan in the scenario.

File: shopfloor/services/delivery.py [hunk 5]
```diff
@@ -130,7 +136,9 @@ def scan_deliver(self, barcode, picking_id=None):
                 return self._deliver_package(picking, package)
 
         if not barcode_valid:
-            product = search.product_from_scan(barcode)
+            product = search.product_from_scan(
+                barcode, use_packaging=(not allow_prepackaged_product)
+            )
             if product:
                 return self._deliver_product(picking, product)
```
Comment (jbaudoux) [inline shopfloor/services/delivery.py:143]: The delivered quantity must always be 1. Same for lot.
Usually, when the package contains a single unit, it is the same barcode on the product and package.

File: shopfloor/services/delivery.py [hunk 6]
```diff
@@ -206,24 +234,37 @@ def _lines_from_lot_domain(self, lot, no_qty_done=True):
...
         if product.tracking in ("lot", "serial"):
             return self._response_for_deliver(
                 picking, message=self.msg_store.scan_lot_on_product_tracked_by_lot()
             )
 
+        limit = 1 if packaging_qty else None
         lines = self.env["stock.move.line"].search(
-            self._lines_from_product_domain(product)
+            self._lines_from_product_domain(product, packaging_qty=packaging_qty),
+            limit=limit,
```
Comment (jbaudoux) [inline shopfloor/services/delivery.py:268]: It is missing a sort by "picking scheduled date, picking id".

**Review by jbaudoux — CHANGES_REQUESTED** [review]

**Review by ivantodorovich — APPROVED** [review]

File: shopfloor/actions/search.py (hunk 4 above)
Comment (jbaudoux) [reply shopfloor/actions/search.py:64]: I remove my comment, it's better to match a product through its barcode packaging even if we don't know the quantity

File: shopfloor/services/delivery.py (hunk 5 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:143]: You mean we should do the call like this?
```suggestion
                return self._deliver_product(picking, product, packaging_qty=1)
```

File: shopfloor/services/delivery.py [hunk 7]
```diff
@@ -206,24 +234,37 @@ def _lines_from_lot_domain(self, lot, no_qty_done=True):
             [self._lines_base_domain(no_qty_done), [("lot_id", "=", lot.id)]]
         )
 
-    def _lines_from_product_domain(self, product, no_qty_done=True):
-        return expression.AND(
+    def _lines_from_product_domain(self, product, no_qty_done=True, packaging_qty=None):
```
Comment (jbaudoux) [inline shopfloor/services/delivery.py]: Similar code is done in https://github.com/OCA/wms/pull/410/files#diff-0ba2c9ccb67d9c1b060b30e26ff7ba99ef0b4733a3c5a6bfeebe64677cc60417R226
Can this be moved outside of the scenario as an action as this is useful in several scenario
cc @mmequignon

File: shopfloor/services/delivery.py (hunk 5 above)
Comment (jbaudoux) [reply shopfloor/services/delivery.py:143]: yes but I would call it product_qty

File: shopfloor/services/delivery.py (hunk 6 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:268]: This means storing the `scheduled_date` field on the move line... I'll check.

File: shopfloor/services/delivery.py (hunk 5 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:143]: it is breaking existing tests, I'm doing something wrong probably.

File: shopfloor/services/delivery.py (hunk 6 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:268]: We now store the `move_id.date` field at the line level in a new `date_planned` (filled by a migration script for upcoming releases).

File: shopfloor/services/delivery.py (hunk 5 above)
Comment (sebalix) [reply shopfloor/services/delivery.py:143]: tests have to be adapted for that use case.

Comment (sebalix) [reply shopfloor/services/delivery.py:143]: This require some changes, I prefer to postpone that in another PR for later (draft PR created here: https://github.com/OCA/wms/pull/418 )

File: shopfloor/services/delivery.py (hunk 7 above)
Comment (sebalix) [reply shopfloor/services/delivery.py]: We will have to postpone this for later, probably addressed when we will write new scenarios to refactor a bit this code.

**Review by jbaudoux — APPROVED** [review]

**Pull request merged**